src/NATS.Client.KeyValueStore/Internal/NatsKVWatcher.cs

```diff
@@ -0,0 +1,401 @@
+                        }
+                        else
+                        {
+                            _logger.LogError("Internal error: unexpected command {Command}", subCommand);
+                        }
+                    }
+                    catch (Exception e)
+                    {
+                        _logger.LogWarning(e, "Command error");
+                    }
+                }
+            }
+        }
+        catch (OperationCanceledException)
+        {
+        }
+        catch (Exception e)
+        {
+            _logger.LogError(e, "Unexpected command loop error");
+        }
+    }
+
+    private async Task ConsumerCreateLoop()
+    {
+        try
+        {
+            while (await _consumerCreateChannel.Reader.WaitToReadAsync(_cancellationToken))
+            {
+                while (_consumerCreateChannel.Reader.TryRead(out var origin))
+                {
+                    try
+                    {
+                        await CreatePushConsumer(origin);
+                    }
+                    catch (Exception e)
+                    {
+                        _logger.LogWarning(e, "Consumer create error");
+                    }
+                }
+            }
+        }
+        catch (OperationCanceledException)
+        {
+        }
+        catch (Exception e)
+        {
+            _logger.LogError(e, "Unexpected consumer create loop error");
+        }
+    }
+
+    private async ValueTask CreatePushConsumer(string origin)
+    {
+        if (_debug)
+        {
+            _logger.LogDebug(NatsKVLogEvents.NewConsumer, "Creating new consumer {Consumer} from {Origin}", Consumer, origin);
+        }
+
+        if (_sub != null)
+        {
+            if (_debug)
+            {
+                _logger.LogDebug(NatsKVLogEvents.DeleteOldDeliverySubject, "Deleting old delivery subject {Subject}", _sub.Subject);
+            }
+
+            await _sub.UnsubscribeAsync();
+            await _sub.DisposeAsync();
+        }
+
+        _sub = new NatsKVWatchSub<T>(_context, _commandChannel, _subOpts, _cancellationToken);
+        await _context.Connection.SubAsync(_sub, _cancellationToken).ConfigureAwait(false);
+
+        if (_debug)
+        {
+            _logger.LogDebug(NatsKVLogEvents.NewDeliverySubject, "New delivery subject {Subject}", _sub.Subject);
+        }
+
+        Interlocked.Exchange(ref _sequenceConsumer, 0);
+
+        var sequence = Volatile.Read(ref _sequenceStream);
+
+        var config = new ConsumerConfiguration
+        {
+            Name = Consumer,
+            DeliverPolicy = ConsumerConfigurationDeliverPolicy.all,
+            AckPolicy = ConsumerConfigurationAckPolicy.none,
+            DeliverSubject = _sub.Subject,
+            FilterSubject = _filter,
+            FlowControl = true,
+            IdleHeartbeat = _idleHbNanos,
+            AckWait = _ackWaitNanos,
+            MaxDeliver = 1,
+            MemStorage = true,
```

Comment:
`MemStorage` may be an issue - some server configurations don't have Memory Store enabled, which will make this fail.  Would be best to leave this out and default to the Stream's backing store

Reply:
AFAIK this is explicitly set by [Go](https://github.com/nats-io/nats.go/blob/main/js.go#L1567) (also Rust I think) clients and in [Ordered Consumer ADR](https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-17.md#consumer-configuration-checks) I believe. I don't know the reason. @piotrpio?

Reply:
Maybe `MemStorage` still works for consumers, even if `max_memory_store` is set to 0 then.  Otherwise this would always fail on NGS, and I am assuming if Go is doing it then someone is successfully using this with NGS :smile: